driver/modern_bpf/add_new_syscalls.md

```diff
@@ -0,0 +1,530 @@
+
+#### 2. Implement the BPF program
+
+As we said we are using the `ringbuf` approach so we can advantage of the `ringbuf` template 👇.
+
+> __Note__: this template can be used both for enter and exit events.
+
+```c
+/* if we are in the enter event */
+SEC("tp_btf/sys_enter") 
+int BPF_PROG(<your_syscall_name>_e,
+      struct pt_regs *regs,
+      long id)
+/* if we are in the exit event*/
+SEC("tp_btf/sys_exit")
+int BPF_PROG(<your_syscall_name>_x,
+      struct pt_regs *regs,
+      long ret)
+/**/
+{
+ struct ringbuf_struct ringbuf;
+ /* Reserve the space for the event in the ring buffer.
+  * Here you should put the dimension we computed in step 1.
+  */
+ if(!ringbuf__reserve_space(&ringbuf, <the_size_of_your_event>))
+ {
+  return 0;
+ }
+
+ /* Fill the event header with your event type */
+ ringbuf__store_event_header(&ringbuf, <your_event_type>);
+
+ /*=============================== COLLECT PARAMETERS  ===========================*/
+
+ /* The params we need to collect. This logic is syscall-specific, we will see it. */
+
+ /*=============================== COLLECT PARAMETERS  ===========================*/
+
+ /* Push the whole event to userspace */
+ ringbuf__submit_event(&ringbuf);
+
+ return 0;
+}
+```
+
+Quite easy, isn't it? Let's fill in the template for our `access` syscall.
+
+```c
+SEC("tp_btf/sys_enter")
+int BPF_PROG(access_e,
+      struct pt_regs *regs,
+      long id)
+{
+ struct ringbuf_struct ringbuf;
+ if(!ringbuf__reserve_space(&ringbuf, ACCESS_E_SIZE))
+ {
+  return 0;
+ }
+
+ ringbuf__store_event_header(&ringbuf, PPME_SYSCALL_ACCESS_E);
+
+ /*=============================== COLLECT PARAMETERS  ===========================*/
+
+ /* As we saw before the `PPME_SYSCALL_ACCESS_E` event sends just one param
+  * to userspace, usually we add a comment for every param with the name and
+  * the type.
+  */
+
+ /* Parameter 1: mode (type: PT_UINT32) */
+
+ /* Now looking at the `access` manual page https://man7.org/linux/man-pages/man2/access.2.html
+  * we can notice that `mode` is the second param (so index `1`):
+  * 
+  *                     0               1
+  * `int access(const char *pathname, int mode)`
+  *
+  * We can use the `extract__syscall_argument` helper to take a syscall argument!
+  * To choose the variable type (`int`) we usally look at the syscall prototype,
+  * so in this case, we have `int mode.`
+  */
+ int mode = extract__syscall_argument(regs, 1);
+
+ /* We saved in our program stack the `mode` argument so we only need to store it into the ringbuffer.
+  * For doing this we have to choose the right ringbuf helper! Do you remember our comment?
+  *
+  * Parameter 1: mode (type: PT_UINT32)
+  *
+  * Well, we need to push a `PT_UINT32`, so as the documentation says
+  * (https://github.com/falcosecurity/libs/blob/master/driver/modern_bpf/helpers/store/ringbuf_store_params.h#L234-L248)
+  * we need to use the `ringbuf__store_u32` helper.
+  */
+ ringbuf__store_u32(&ringbuf, access_flags_to_scap(mode));
```

Comment:
It would be worth to have note on `xx_flags_to_scap` in sense of abstracting flags for the userspace libs.

Reply:
sure I will add it, ty!